graphql_compiler/interpreter/typedefs.py

```diff
@@ -0,0 +1,208 @@
+    However, for the sake of easier debugging and testing using the built-in functionality in
+    this library, it is desirable to make DataToken be a deep-copyable type that implements
+    equality beyond a simple referential equality check.
+
+    ## The InterpreterAdapter API
+
+    The methods in the InterpreterAdapter API are all designed to support generator-style operation,
+    where data is produced and consumed only when required. Here is a high-level description of
+    the methods in the InterpreterAdapter API:
+    - get_tokens_of_type() produces an iterable of DataTokens of the type specified by its argument.
+      The calling function will wrap the DataTokens into a bookkeeping object called a DataContext,
+      where a particular token is currently active and specified in the "current_token" attribute.
+    - For an iterable of such DataContexts, project_property() can be used to get the value
+      of one of the properties on the vertex type represented by the currently active DataToken
+      in each DataContext; project_property() therefore returns an iterable of
+      tuples (data_context, value).
+    - project_neighbors() is similar: for an iterable of DataContexts and a specific edge name,
+      it returns an iterable (data_context, iterable_of_neighbor_tokens) where each result
+      contains the DataTokens of the neighboring vertices along that edge for the vertex whose
+      DataToken is currently active in that DataContext.
+    - can_coerce_to_type() is used to check whether a DataToken corresponding to one vertex type
+      can be safely converted into one representing a different vertex type. Given an iterable of
+      DataContexts and the name of the type to which the conversion is attempted, it produces
+      an iterable of tuples (data_context, can_coerce) where can_coerce is a boolean.
+
+    ## Performance and optimization opportunities
+
+    The design of the API, including its generator-style operation, enable a variety of
+    optimizations to either happen automatically or be available with minimal additional work.
+    A few simple examples:
+    - Interpreters perform lazy evaluation by default: if exactly 3 query results are requested,
+      then only the minimal data necessary to produce *exactly 3 rows' worth* of outputs is loaded.
+    - When computing a particular result, data loading for output fields is deferred
+      until *after* all filtering operations have been completed, to minimize data loads.
+    - Data caching is easy to implement within this API -- simply have
+      your API function's implementation consult a cache before performing the requested operation.
+    - Batch-loading of data can be performed by simply advancing the input generator multiple times,
+      then operating on an entire batch of input data before producing corresponding outputs:
+
+        def project_property(
```

Comment:
Yay! 🙌  I added it because I found through talking to people that this pattern seems to be in the "obvious in retrospect" category, and it's a very easy and powerful way to make queries faster since most APIs support some flavor of batch operation.